compiler/rustc_hir_analysis/src/check/wfcheck.rs

```diff
@@ -1712,7 +1712,15 @@ fn receiver_is_valid<'tcx>(
     let cause =
         ObligationCause::new(span, wfcx.body_def_id, traits::ObligationCauseCode::MethodReceiver);
 
-    let can_eq_self = |ty| infcx.can_eq(wfcx.param_env, self_ty, ty);
+    let can_eq_self = |ty| {
+        wfcx.infcx.probe(|_| {
+            let ocx = ObligationCtxt::new(wfcx.infcx);
+            let Ok(()) = ocx.eq(&ObligationCause::dummy(), wfcx.param_env, self_ty, ty) else {
+                return false;
+            };
+            ocx.select_where_possible().is_empty()
+        })
+    };
```

Comment:
does `wfcx.eq` ever fail when `can_eq_self` returns true? I feel like it can#t and all this code should just use `commit_if_ok` :thinking: could be a followup

Reply:
Ok, I just did it in this PR. You should probably read it and convince yourself it's correct, since I did kinda simplify a lot of the control flow which was redundant.